packages/st2/rpm/st2.spec

```
@@ -25,6 +25,24 @@ BuildRequires: python3-devel
 BuildRequires: python3-setuptools
 %endif  # Requires for RHEL 8
 
+%if 0%{?rhel} >= 7
+# By default on EL 8, RPM helper scripts will try to generate Requires: section which lists every
+# Python dependencies. That process / script works by recursively scanning all the package Python
+# dependencies which is very slow (5-6 minutes).
+# Our package bundles virtualenv with all the dependendencies and doesn't rely on this metadata
+# so we skip that step to vastly speed up the build.
+# Technically we also don't Require or Provide any of those libraries auto-detected by that script
+# because those are only used internally inside a package specific virtual environment.
+# Same step also does not run on EL7.
+# See https://github.com/StackStorm/st2-packages/pull/697#issuecomment-808971874 and that PR for
+# more details.
+# That issue was found by enabling rpmbuild -vv flag.
+%undefine __pythondist_provides
+%undefine __pythondist_requires
+%undefine __python_provides
+%undefine __python_requires
+%endif
```

I'm fine with the `Provides` change, that's a great cleanup, especially if it improves the build timing! :+1: 

I'm more worried about the Requires section. If it doesn't touch the Requires, do we really need to undefine those?

```suggestion
%undefine __pythondist_provides
%undefine __python_provides
%endif
```

---

I don't think it should hurt anything by having it there since we don't explicitly depend on any Python dependency via yum package (we install those manually inside virtualenv).

In fact, I would argue it's probably better to have it there to detect any issues where that macro may declare dependencies which we don't want by doing "magic introspection of all the Python deps" (similar to what happened with ``Provides`` and went undetected for so long).

In short, it only affects Python dependencies and if it indeed breaks something - our end to end tests should be there to catch that. If they don't, well, we should fix that.

---

I'm not convinced of changing the behavior for `Requires` and trying to catch that via e2e tests later.

Did `%undefine __pythondist_requires` and `%undefine __python_requires` affect the build timing or was it only `Provides` change that really helped with the speedup?

---

Yes, both of the changes help.

First I accidentally only added ``requires`` and the build was faster, but not as much as when also adding provides override.

And I think end to end tests should be there exactly for reasons like that - to catch any kind of "end to end" issues with the packages. Including invalid package metadata or similar.

Technically, we could also add some quick assertions to this repo, but this could be a slippery slope since you could argue end to end tests are a better fit for catching issues like that.

---

But again, if you feel strongly about it, I'm fine with adding some quick "grep based" assertions to the build job which verifies that some specific / minimum dependencies are defined as part of "Requires".

---

Alright,
Let me check the timing including the `Provides` and excluding the `Requires` change to understand your point and tradeoffs better.

---

Looks good to me, approved it. Thanks for research.